### ISSUE TITLE
Cloud API firmware 2.x.x

### ISSUE THREAD

**Issue by koalazak** [koalazak/dorita980#25]
Hi guys,
I have almost everything ready to make the Cloud API possible:
- cloud discovery
- cloud aws login
- suscribe to topics in the cloud to receive state updates
- make changes in the state to set preferences ( schedule, bust, etc)

But i dont found the correct `topic` and `message` content to send basic commands like `start` or `stop`.
Anybody sniff that data? or *has* that data?

My sniff data is weired and malformed, i dont know if my sslsplit is showingme the info in the correct encoding. When I send a command with my phone over the cloud I see some bytes in the comunication but no one string like topic o json message.

can anybody help?

here is the working snippet:

```
const AWSIoTData = require('aws-iot-device-sdk');
const AWS = require('aws-sdk');
const request = require('request-promise');
// install with: npm install aws-iot-device-sdk aws-sdk request-promise request

const ROBOT_BLID = ''; // same as local api
const ROBOT_PASSWORD = ''; // same as local api
const APP_ID = ''; // like IOS-12345678-1234-1234-1234-123456789098

function cloudDiscovery () {
  var requestOptions = {
    'method': 'GET',
    'uri': `https://disc-prod.iot.irobotapi.com/v1/robot/discover/${ROBOT_BLID}`,
    'json': true
  };
  return request(requestOptions);
}

function cloudLogin () {
  return cloudDiscovery().then(function (discoveryData) {
    var postData = {
      'associations': {
        '0': {
          'robot_id': ROBOT_BLID,
          'deleted': false,
          'password': ROBOT_PASSWORD
        }
      },
      'app_id': APP_ID
    };

    var requestOptions = {
      'method': 'POST',
      'headers': {
        'Content-Type': 'application/json',
        'User-Agent': 'aspen/1.9.1.184.1 CFNetwork/808.2.16 Darwin/16.3.0'
      },
      'uri': `${discoveryData.httpBase}/v1/login`,
      'body': postData,
      'json': true
    };

    return request(requestOptions).then((rawLoginResponse) => {
      return {login: rawLoginResponse.associations['0'], credentials: rawLoginResponse.credentials, discovery: discoveryData};
    });
  });
}

function initMQTT (amazonData) {
  var awsConfiguration = {
    poolId: amazonData.credentials.CognitoId,
    region: amazonData.discovery.awsRegion
  };

  var AWSConfiguration = awsConfiguration;

  var clientId = ROBOT_BLID + '-' + (Math.floor((Math.random() * 100000) + 1));

  AWS.config.region = AWSConfiguration.region;

  AWS.config.credentials = new AWS.CognitoIdentityCredentials({
    IdentityPoolId: AWSConfiguration.poolId
  });

  const mqttClient = AWSIoTData.device({
    region: AWS.config.region,
    clientId: clientId,
    protocol: 'wss',
    maximumReconnectTimeMs: 8000,
    debug: true,
    accessKeyId: amazonData.credentials.AccessKeyId,
    secretKey: amazonData.credentials.SecretKey,
    sessionToken: amazonData.credentials.SessionToken
  });

  mqttClient.on('connect', function (e) {
    console.log('connect!', e);

    mqttClient.subscribe('$aws/things/' + ROBOT_BLID + '/shadow/#'); // all subtopics

    // const cmd = {'state': {'desired': {'cleanSchedule': {'cycle': ['none', 'none', 'none', 'none', 'none', 'none', 'none'], 'h': [17, 10, 10, 12, 10, 13, 17], 'm': [0, 30, 30, 0, 30, 30, 0]}}}};
    // mqttClient.publish('$aws/things/' + ROBOT_BLID + '/shadow/update', JSON.stringify(cmd));
  });

  mqttClient.on('reconnect', function () {
    console.log('reconnect!');
  });

  mqttClient.on('message', function (t, m) {
    console.log('message:');
    console.log('topic:', t);
    console.log(m.toString());
  });

  mqttClient.on('delta', function (m) {
    console.log('delta:');
    console.log(m);
  });
  mqttClient.on('status', function (m) {
    console.log('status:');
    console.log(m);
  });

  mqttClient.on('data', function (m) {
    console.log('data:');
    console.log(m);
  });

  mqttClient.on('error', function (e) {
    console.log('error:');
    console.log(e);
  });

  mqttClient.on('packetreceive', function (m) {
    console.log('packetreceive:');
    console.log(m);
  });
}

cloudLogin().then((credentialData) => {
  console.log(credentialData);
  initMQTT(credentialData);
}).catch(console.log);
```

**Comment by koalazak** [comment]
@akpotter @Thoro :)

**Comment by letier** [comment]
Hey @koalazak!

I'd like to help. Can you provide some instructions?

How do I find the app id to run your snippet? I guess it's from the mobile app?

Cheers

**Comment by koalazak** [comment]
I think you can use any string with that format. I get my app ID sniffing the trafic.

**Comment by letier** [comment]
OK, it seems to do something. Which is the relevant output? :)

**Comment by letier** [comment]
So i did run a start stop cycle through the cloud and got readable output from your script.

Is this what you're looking for?

**Comment by koalazak** [comment]
the output is all the messages sent to the topic `$aws/things/' + ROBOT_BLID + '/shadow` (and subtopics) by the robot or by the official mobile app.
When you say `through the cloud and got readable output` you mean using the official mobile app? or using this script (publishing to a topic with this script)?

What im looking for is what message and what topic we suppose to use in `mqttClient.publish()` method in this script to start/stop the robot via this script.

**Comment by letier** [comment]
Running this script, then triggering the bot via the official mobile app.

When I start the script, the output ends with "undefined".

**Comment by letier** [comment]
One sec, I'm sanitizing the output.

**Comment by koalazak** [comment]
ok, you are looking the messages published by the mobile app or robot to that topics. that is what i get too. and using the `mqttClient.publish()` method I can set new schedule for example. But What im looking for is what message and what topic we suppose to use in mqttClient.publish() method in this script to start/stop the robot via this script.

**Comment by letier** [comment]
So do you have some instructions to obtain them?

I have a ton of JSON output. But only after triggering something via the official mobile app.

**Comment by koalazak** [comment]
you can start making a MITM attack to sniff the data between the official mobile app and the cloud, getting the mqtt packets and decode them. Google about MITM

**Comment by letier** [comment]
No TLS on that?

No chance to reconstruct the command from something like this?
`message:
topic: $aws/things/ROBOT_BLID_HERE/shadow/update/accepted
{"state":{"reported":{"lastCommand":{"command":"start","time":1488399056,"initiator":"rmtApp"}}},"metadata":{"reported":{"lastCommand":{"command":{"timestamp":1488399056},"time":{"timestamp":1488399056},"initiator":{"timestamp":1488399056}}}},"version":1617,"timestamp":1488399057}`

**Comment by koalazak** [comment]
yes, TLS on that. bypassing with `sslsplit`.
I try some variants of that message but no way.

**Comment by letier** [comment]
Is there a way to trigger the cloud api when you're on wifi?

I have a proxy running to monitor the traffic.

**Comment by letier** [comment]
Never mind, i isolated my networks, now it tries to go through the cloud.

I see a message with the app id, i also get the cloud icon in the app. But now my roomba doesn't react when i press start.

**Comment by letier** [comment]
Reset everything, now the roomba reacts, but I don't see the commands in the proxy. Only the requests for mission history, login, etc. Will debug a bit more...

**Comment by koalazak** [comment]
Its like the start/stop commands are in other format not json, like raw mqtt packets, a few bytes...

**Comment by koalazak** [comment]
Make sure you are creating certs with a valid CommonName.
The mobile app is validating the CN field in the roomba cert with the format 'Roomba-{number16}' if it is not valid, then the mobile app disconnect. (the validation is in the mobile app, so you need a selfisgned cert with this CN if you want to perform a sslsplit to sniff the trafic

**Comment by koalazak** [comment]
but if you are seeing the mission commands and commands when you set preferences, there is no problem with the cert.

**Comment by letier** [comment]
Hmm, I used mitmproxy instead of sslsplit. I'm a bit too tired to read through how to use sslsplit right now. 

There should be a option to see raw tcp traffic in mitmproxy. I'm seeing the http traffic, but I'm guessing mqtt gets lost.

**Comment by letier** [comment]
I think i have a working setup now redirecting to sslsplit. A test port 80 request got logged. I tried redirecting 8883 but get no traffic. Any tips to debug the problem?

**Comment by letier** [comment]
OK, finally got a start message :)

**Comment by koalazak** [comment]
yes sure.
I got the message too. But I cant figure out how to reproduce it in my test snippet. Can you?

**Comment by letier** [comment]
I got to a version mismatch. It seems to be a sequence number. The question is where from.

**Comment by koalazak** [comment]
well, that is a step forward! There is not in state object?

**Comment by koalazak** [comment]
version is in every message:
```
{
"state":{"reported":{"svcEndpoints":{"svcDeplId":"v007"}}},
"metadata":{"reported":{"svcEndpoints":{"svcDeplId":{"timestamp":1488553359}}}},
"version":3345,
"timestamp":1488553359
}
```

you can parse all the messages and store the version to use in the call.

can you share your code? zaktu.x@gmail.com

**Comment by letier** [comment]
Will send it later

**Comment by letier** [comment]
Sent the log.

Btw, sending the previous command without the version actually gets a valid reply. But the robot doesn't start.

The message I sent was:
`{"state":{"reported":{"lastCommand":{"command":"start","time":1488399056,"initiator":"rmtApp"}}},"metadata":{"reported":{"lastCommand":{"command":{"timestamp":1488399056},"time":{"timestamp":1488399056},"initiator":{"timestamp":1488399056}}}},"timestamp":1488399057}`

**Comment by letier** [comment]
`{
      "state":{
        "desired":{
          "command":{
            "command":"start","time":+new Date(),"initiator":"rmtApp"
          }
        }
      },
      "timestamp":+new Date()
    }`

Bam! Your working start command! :)

**Comment by koalazak** [comment]
I think that string you see in logs are a `post-execution` message sent by the robot after receibe the command. And the real command is all the ugly bytes before that string

**Comment by koalazak** [comment]
nice!
I tried something like that before but nothing. Maybe I dont use the `time` filed in my tests...
Testing now....

**Comment by letier** [comment]
I actually got the robot to react with the last posted one. "start" for start and "stop" for stop.

There seems to be some kind of problems with the timestamps though, as it will start over and over again.

**Comment by letier** [comment]
Yes, it seems to be important.

**Comment by koalazak** [comment]
oh crap, that works but the robot now is receiving the start command every second :p.
let me kwno if you can stop it :p

**Comment by letier** [comment]
Send the stop message. Same format. x)

**Comment by letier** [comment]
I can't, but my hope is that a hard reset will be able to fix it. :)

Let me try if the app is still able to start it.

**Comment by letier** [comment]
Actually it is in a start stop loop. I'm resetting mine.

**Comment by letier** [comment]
Yep, a reset fixed it. But that means we have to find out where to get the right timestamps.

**Comment by letier** [comment]
Or maybe it's not about the timestamps but some of the other parts of the messages. The messages in the app seem to contain more information.

**Comment by koalazak** [comment]
reset as 10 second the start button?

**Comment by letier** [comment]
Didn't try that. I used the "reset roomba" option in the app. But you have to pair it again.

**Comment by koalazak** [comment]
reset with 10 second start button doest work. but yes with the App.

**Comment by iosdeveloper** [comment]
New firmware is rolling out (again). Does anybody know what's changed? Hopefully nothing broke.
![img_4182](https://cloud.githubusercontent.com/assets/655921/23582110/5c79d218-0123-11e7-9834-654de003d918.JPG)

**Comment by koalazak** [comment]
getpassword works, state events and cleaning preferes too.
commands like start/stop/dock doesnt work :/
next week i will try to fix

**Comment by koalazak** [comment]
soundVer: '31'
uiSwVer: '4582'
navSwVer: '01.09.09' >> '01.11.02'
wifiSwVer: '20902' >> '20923'
mobilityVer: '5309' >> '5420'
bootloaderVer: '3580',
umiVer: '5' >> '6'
softwareVer: 'v2.0.0-34' >> 'v2.2.5-2'

**Comment by koalazak** [comment]
and a new creepy property in the state: `mapUploadAllowed: true`

**Comment by koalazak** [comment]
@iosdeveloper opened a new issue for that: #31

**Comment by koalazak** [comment]
@letier do you have any progress in the cloud api reverse engeniering?

**Comment by letier** [comment]
No luck. :(

It works with the iOS app at the moment, but doesn't do anything when I send the signals. The reset option in the app is greyed out, so I have no clue if I got it stuck in a strange state again.

**Comment by letier** [comment]
Managed to reset and catch another conversation with the app. Saw the start message. But I'm suddenly not getting it to start anymore. And there seems to be all kind of state from my past trials saved. I'll need to find a way to clean it up. Posting via the snippet doesn't seem to work anymore.

**Comment by letier** [comment]
So the situation seems to be that you somehow have to switch between certain states.

The following cycles works for me:
{ "state":{"desired":{"command":{"command":"start","time":Math.floor(new Date()/1000),"initiator":"rmtApp"}}}}
{ "state":{"desired":{"command":{"command":"clean","time":Math.floor(new Date()/1000),"initiator":"rmtApp"}}}}
{ "state":{"desired":{"command":null}}}

{ "state":{"desired":{"command":{"command":"stop","time":Math.floor(new Date()/1000),"initiator":"rmtApp"}}}}
{ "state":{"desired":{"command":null}}}

I'm not sure if clean is a real command. I just tried it and the robot started moving.

**Comment by koalazak** [comment]
I dont like that aprouch :p may there is another fancy way haha. If i dont foudn the way y just implement that in next version.

**Comment by andrey3diq** [comment]
Hi Guys, did you try to contact iRobot support to get some additional info?

**Comment by pschmitt** [comment]
I dont believe they'd be of much help:
https://homesupport.irobot.com/app/answers/detail/a_id/9840

**Comment by andrey3diq** [comment]
Snippet doesn't work for me.
`body:
 { errorMessage: 'Authentication failed',
   errorType: 'AspenError.AuthenticationFailed' }`

I think it can be because I use fake APP_ID. Could you please share the correct one?

**Comment by koalazak** [comment]
Nice, are you trying to send commands with `mqttClient.publish()` ?

**Comment by andrey3diq** [comment]
I use AWSIoTData.thingShadow instead of AWSIoTData.device to create the object.
This is my code:

```
    const mqttClient = AWSIoTData.thingShadow(connectionOptions);
    var clientTokenUpdate;

    mqttClient.on('connect', function (e) {
        console.log('connect!', e);

        mqttClient.register(ROBOT_BLID, {debug: true}, function () {
            var startCmd = {command: {command: 'start'}};
            // this is to reset command
            //var startCmd = null;
            clientTokenUpdate = mqttClient.update(ROBOT_BLID, {'state': {'desired': startCmd}});

            // this is to get the full thing shadow state
            //clientTokenUpdate = mqttClient.get(ROBOT_BLID);
            if (clientTokenUpdate === null) {
                console.log('update shadow failed, operation still in progress');
            }
        });
    });

    mqttClient.on('status',
        function (thingName, stat, clientToken, stateObject) {
            console.log('received ' + stat + ' on ' + thingName + ': ' +
                JSON.stringify(stateObject));

    mqttClient.on('delta',
        function (thingName, stateObject) {
            console.log('received delta on ' + thingName + ': ' +
                JSON.stringify(stateObject));
        });

    mqttClient.on('timeout',
        function (thingName, clientToken) {
            console.log('received timeout on ' + thingName +
                ' with token: ' + clientToken);

```

When I set the 'desired' state to 'command' it doesn't appear in the shadow state but instead it acts on my device. 
It doesn't clear the 'desired' state and I think this is why it's looping constantly trying to set this property.
The same behavior was when I tried to set the boolean property to string value. It keeps this property in 'desired' section and constantly sends the 'delta' events. You can check if you get full state.
I think we should clean the 'desired' command manually. Right after it acts.

**Comment by aimproxy** [comment]
Hi, I'm implementing ur library in C for an IoT School Project, https://github.com/roombavacuum/libroomba
and when I sent a command to my roomba e5 v3 in local, I got this:
```
MQTT Message: Topic wifistat, Qos 0, Len 163
Payload (0 - 80): {"state":{"reported":{"netinfo":{"dhcp":true,"addr":3232235870,"mask":4294967040
MQTT Message: Done
MQTT Message: Topic wifistat, Qos 0, Len 68
Payload (0 - 68): {"state":{"reported":{"wifistat":{"wifi":1,"uap":false,"cloud":1}}}}
MQTT Message: Done
MQTT Message: Topic wifistat, Qos 0, Len 163
Payload (0 - 80): {"state":{"reported":{"netinfo":{"dhcp":true,"addr":3232235870,"mask":4294967040
MQTT Message: Done
MQTT Message: Topic wifistat, Qos 0, Len 82
Payload (0 - 80): {"state":{"reported":{"wlcfg":{"sec":7,"ssid":"566F6461666F6E652D304134413632"}}
MQTT Message: Done
MQTT Message: Topic wifistat, Qos 0, Len 50
Payload (0 - 50): {"state":{"reported":{"mac":"d0:c5:d3:bd:21:31"}}}
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 40
Payload (0 - 40): {"state":{"reported":{"country": "PT"}}}
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 43
Payload (0 - 43): {"state":{"reported":{"cloudEnv": "prod"}}}
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 61
Payload (0 - 61): {"state":{"reported":{"svcEndpoints":{"svcDeplId": "v011"}}}}
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 37
Payload (0 - 37): {"state":{"reported":{"name": "e5"}}}
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 470
Payload (0 - 80): {"state":{"reported":{"lastDisconnect":4,"cap":{"ota":1,"eco":1,"svcConf":1},"ba
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 378
Payload (0 - 80): {"state":{"reported":{"bbmssn":{"nMssn":254,"nMssnOK":219,"nMssnF":35,"aMssnM":1
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 475
Payload (0 - 80): {"state":{"reported":{"cleanSchedule":{"cycle":["none","none","none","none","non
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 460
Payload (0 - 80): {"state":{"reported":{"langs":[{"en-US":0},{"en-GB":15},{"fr-FR":1},{"de-DE":2},
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 244
Payload (0 - 80): {"state":{"reported":{"tz":{"ver":7,"events":[{"dt":1564675200,"off":60},{"dt":1
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 294
Payload (0 - 80): {"state":{"reported":{"lastCommand":{
   "command": "start",
   "time": 15785928
MQTT Message: Done
MQTT Message: Topic $aws/things/3178480822035600/shadow/update, Qos 0, Len 201
Payload (0 - 80): {"state":{"reported":{"cleanMissionStatus":{"cycle":"clean","phase":"run","expir
MQTT Message: Done
MQTT Message: Topic wifistat, Qos 0, Len 55
Payload (0 - 55): {"state":{"reported":{"signal":{"rssi":-67,"snr":23}}}}
```

Is that helpful to ur could implementation?

**Comment by koalazak** [comment]
will close this issue and reopen if start with the research again

**Comment by machineonamission** [comment]
OH i know how to work with these! this is the AWS device shadow

https://docs.aws.amazon.com/iot/latest/developerguide/device-shadow-mqtt.html

the MQTT endpoints are documented, though how much you're able to control via mqtt alone is not certain

but perhaps we can hit endpoints like /get. i will give it a try

**Comment by machineonamission** [comment]
oh

<img width="751" height="673" alt="Image" src="https://github.com/user-attachments/assets/928f4999-4190-4408-8bdb-fa8dd6a9a458" />

**Comment by machineonamission** [comment]
alright got MITM set up to sniff irobot app and cloud

so the /shadow/get/accepted payload from server -> client sends the entire shadow/mqtt schema

for my m6:
```json
{"state":{"reported":{"nsmip":2,"svcEndpoints":{"svcDeplId":"v011"},"sku":"m611220","cap":{"oMode":4,"edge":0,"maps":3,"pmaps":10,"mc":2,"tHold":1,"tLine":2,"area":1,"eco":1,"multiPass":2,"team":1,"pp":0,"pose":2,"lang":2,"5ghz":1,"prov":3,"sched":2,"svcConf":1,"ota":2,"log":2,"langOta":0,"ns":1,"tileScan":1},"digiCap":{"appVer":0}}},"metadata":{"reported":{"nsmip":{"timestamp":1774126877},"svcEndpoints":{"svcDeplId":{"timestamp":1774302927}},"sku":{"timestamp":1774302927},"cap":{"oMode":{"timestamp":1774302927},"edge":{"timestamp":1774302927},"maps":{"timestamp":1774302927},"pmaps":{"timestamp":1774302927},"mc":{"timestamp":1774302927},"tHold":{"timestamp":1774302927},"tLine":{"timestamp":1774302927},"area":{"timestamp":1774302927},"eco":{"timestamp":1774302927},"multiPass":{"timestamp":1774302927},"team":{"timestamp":1774302927},"pp":{"timestamp":1774302927},"pose":{"timestamp":1774302927},"lang":{"timestamp":1774302927},"5ghz":{"timestamp":1774302927},"prov":{"timestamp":1774302927},"sched":{"timestamp":1774302927},"svcConf":{"timestamp":1774302927},"ota":{"timestamp":1774302927},"log":{"timestamp":1774302927},"langOta":{"timestamp":1774302927},"ns":{"timestamp":1774302927},"tileScan":{"timestamp":1774302927}},"digiCap":{"appVer":{"timestamp":1774213528}}}},"version":13548,"timestamp":1774329488}
```
i7:
```json
{"state":{"reported":{"nsmip":2,"svcEndpoints":{"svcDeplId":"v011"},"sku":"i710020","cap":{"binFullDetect":2,"addOnHw":1,"oMode":2,"dockComm":1,"wDevLoc":2,"bleDevLoc":1,"maps":3,"pmaps":10,"mc":2,"tLine":2,"area":1,"eco":1,"multiPass":2,"edge":0,"pp":0,"team":1,"pose":2,"lang":2,"5ghz":1,"prov":3,"sched":2,"svcConf":1,"ota":2,"log":2,"langOta":0,"ns":1,"tileScan":1},"digiCap":{"appVer":0}}},"metadata":{"reported":{"nsmip":{"timestamp":1756218600},"svcEndpoints":{"svcDeplId":{"timestamp":1774307384}},"sku":{"timestamp":1774307384},"cap":{"binFullDetect":{"timestamp":1774307384},"addOnHw":{"timestamp":1774307384},"oMode":{"timestamp":1774307384},"dockComm":{"timestamp":1774307384},"wDevLoc":{"timestamp":1774307384},"bleDevLoc":{"timestamp":1774307384},"maps":{"timestamp":1774307384},"pmaps":{"timestamp":1774307384},"mc":{"timestamp":1774307384},"tLine":{"timestamp":1774307384},"area":{"timestamp":1774307384},"eco":{"timestamp":1774307384},"multiPass":{"timestamp":1774307384},"edge":{"timestamp":1774307384},"pp":{"timestamp":1774307384},"team":{"timestamp":1774307384},"pose":{"timestamp":1774307384},"lang":{"timestamp":1774307384},"5ghz":{"timestamp":1774307384},"prov":{"timestamp":1774307384},"sched":{"timestamp":1774307384},"svcConf":{"timestamp":1774307384},"ota":{"timestamp":1774307384},"log":{"timestamp":1774307384},"langOta":{"timestamp":1774307384},"ns":{"timestamp":1774307384},"tileScan":{"timestamp":1774307384}},"digiCap":{"appVer":{"timestamp":1756303539}}}},"version":174460,"timestamp":1774329488}
```

my j9:
```json
{"state":{"reported":{"digiCap":{"appVer":0,"smartClean":1},"nsmip":2,"svcEndpoints":{"svcDeplId":"v011"},"sku":"j955020","cap":{"carpetBoost":1,"binFullDetect":2,"addOnHw":1,"pw":0,"oMode":10,"floorTypeDetect":3,"odoa":7,"dockComm":1,"maps":3,"pmaps":10,"mc":2,"sem2umf":2,"tLine":2,"area":1,"eco":1,"multiPass":2,"pp":0,"team":1,"pose":2,"lang":2,"gentle":1,"hm":0,"rNav":2,"5ghz":1,"prov":3,"sched":2,"svcConf":1,"ota":2,"log":2,"langOta":0,"ns":1,"bleLog":1,"expectingUserConf":2,"idl":1}}},"metadata":{"reported":{"digiCap":{"appVer":{"timestamp":1774126213},"smartClean":{"timestamp":1774126213}},"nsmip":{"timestamp":1774126173},"svcEndpoints":{"svcDeplId":{"timestamp":1774302555}},"sku":{"timestamp":1774302555},"cap":{"carpetBoost":{"timestamp":1774302555},"binFullDetect":{"timestamp":1774302555},"addOnHw":{"timestamp":1774302555},"pw":{"timestamp":1774302555},"oMode":{"timestamp":1774302555},"floorTypeDetect":{"timestamp":1774302555},"odoa":{"timestamp":1774302555},"dockComm":{"timestamp":1774302555},"maps":{"timestamp":1774302555},"pmaps":{"timestamp":1774302555},"mc":{"timestamp":1774302555},"sem2umf":{"timestamp":1774302555},"tLine":{"timestamp":1774302555},"area":{"timestamp":1774302555},"eco":{"timestamp":1774302555},"multiPass":{"timestamp":1774302555},"pp":{"timestamp":1774302555},"team":{"timestamp":1774302555},"pose":{"timestamp":1774302555},"lang":{"timestamp":1774302555},"gentle":{"timestamp":1774302555},"hm":{"timestamp":1774302555},"rNav":{"timestamp":1774302555},"5ghz":{"timestamp":1774302555},"prov":{"timestamp":1774302555},"sched":{"timestamp":1774302555},"svcConf":{"timestamp":1774302555},"ota":{"timestamp":1774302555},"log":{"timestamp":1774302555},"langOta":{"timestamp":1774302555},"ns":{"timestamp":1774302555},"bleLog":{"timestamp":1774302555},"expectingUserConf":{"timestamp":1774302555},"idl":{"timestamp":1774302555}}}},"version":802,"timestamp":1774329488}
```

notably, there are lots of different named shadows.

i observed ```rw-software, rw-constatus, rw-settings, ro-services, ro-stats, ro-configinfo, ro-currentstate```

much of the communication with the roomba is done not through the "$aws/" endpoint bu with the "v011-irbthbu" endpoint. this seems to be how commands are sent (eg PUBLISH to `v011-irbthbu/things/669FA64B995C4A5CB8811DB968FBDE66/cmd` the payload `{"command": "dock", "initiator": "rmtApp", "time": 1774330609}` to dock)

also this is the payload i sent to start a job

```json
{"command": "start", "initiator": "rmtApp", "ordered": 1, "pmap_id": "CtimwQmoTba0pTKC8pabNg", "regions": [{"params": {"carpetBoost": true, "gentleMode": 0, "noAutoPasses": true, "twoPass": false, "vacHigh": false}, "region_id": "3", "type": "rid"}, {"params": {"carpetBoost": true, "gentleMode": 0, "noAutoPasses": true, "twoPass": false, "vacHigh": false}, "region_id": "6", "type": "rid"}, {"params": {"carpetBoost": true, "gentleMode": 0, "noAutoPasses": true, "twoPass": false, "vacHigh": false}, "region_id": "5", "type": "rid"}, {"params": {"carpetBoost": true, "gentleMode": 0, "noAutoPasses": true, "twoPass": false, "vacHigh": false}, "region_id": "7", "type": "rid"}, {"params": {"carpetBoost": true, "gentleMode": 0, "noAutoPasses": true, "twoPass": false, "vacHigh": false}, "region_id": "2", "type": "rid"}, {"params": {"carpetBoost": true, "gentleMode": 0, "noAutoPasses": true, "twoPass": false, "vacHigh": false}, "region_id": "4", "type": "rid"}], "time": 1774330553, "user_pmapv_id": "260321T214337"}
```

it seems you can tell it to do room-by-room cleaning via mqtt, but i didnt see any map *data* being sent over mqtt.

here is the wireshark dissection of my packets just poking around in the app while capturing.

[roomba mqtt.txt](https://github.com/user-attachments/files/26201635/roomba.mqtt.txt)

[roomba.pdml.xml](https://github.com/user-attachments/files/26201661/roomba.pdml.xml)

my setup:
- rooted android phone
  - Always Trust User Certificates (zygisk)
  - SSL Killer (Xposed/LSposed)
- mitmproxy in wireguard mode, outputting keylog
- wireshark capturing pc ethernet
  - if TLS decryption is set up right, you can see the outbound and inbound packets decrypted. not from/to your device (couldnt figure out wireguard) but from your pc to whatever server
- [mqtt over wss parser](https://github.com/frogkind/wireshark-mqtt-websocket-plugin)
  - this + wireshark TLS decrypt means you can just filter for "mqtt"
  - also you can force mqtt packets to decode as json in wireshark
